M2/Macaulay2/packages/MacaulayPosets.m2

```diff
@@ -0,0 +1,1956 @@
+    try R2/I2 then T := R2/I2 else return "Fourth input must be an ideal of the second polynomial ring.";
+    
+    ringTensorProduct := tensor(S, T);
+    
+    -- The ambient ring of the ring tensor product is the same as what we want for the ring fiber product.
+    fiberRing := ambient ringTensorProduct; 
+    fiberIdeal := ideal ringTensorProduct;
+    extraIdeal := 0_fiberRing;
+    extraIdealList := {};
+    
+    -- Get the two maps into the tensor product.
+    -- We assume that the fiberRing gained from the tensor product will be in the correct form with the variables of the 1st ring at the start, and the variables of the 2nd ring at the end.
+    R1toTensor := map(fiberRing, R1, take(gens fiberRing, #(gens R1)));
+    R2toTensor := map(fiberRing, R2, take(gens fiberRing, -#(gens R2)));
+    
+    scan(gens R1, g->scan(gens R2, h->(
+        extraIdealList = append(extraIdealList, R1toTensor(g) * R2toTensor(h));
+    )));
+    
+    extraIdeal = ideal( extraIdealList );
+    fiberIdeal = fiberIdeal + extraIdeal;
+    
+    if connected then (
+        maxs := {0};
+        
+        scan({I1, I2}, I -> (
+            P := getPoset I;
+            
+            maxP := maximalElements P;
+            
+            if not (1 == #maxP) then return "The monomial posets of the rings must have unique maximal elements.";
+            
+            maxs = append(maxs, maxP#0);
+        ));
+        
+        fiberIdeal = fiberIdeal + ideal(R1toTensor(lift(maxs#1,R1)) - R2toTensor(lift(maxs#2,R2)));
+    );
+    
+    fiberRing/fiberIdeal
+    );
+
+-- Since Binary=>true, these four methods should return something of the same type as the inputs.
+ringFiberProduct(Ideal, Ideal) := (I1, I2) -> (
+    ideal ringProductHelper(I1, I2, false)
+);
+ringConnectedSum(Ideal, Ideal) := (I1, I2) -> (
+    ideal ringProductHelper(I1, I2, true)
+);
+
+ringFiberProduct(QuotientRing, QuotientRing) := (S, T) -> ( -- If our input is two quotient rings do: 
+    ringProductHelper(ideal S, ideal T, false)
+);
+ringConnectedSum(QuotientRing, QuotientRing) := (S, T) -> (
+    ringProductHelper(ideal S, ideal T, true)
+);
+
+--------------------------------------------------
+--- Shadows
+--------------------------------------------------
+
+-- If P is a Poset and S is a subset of its ground set P_*, then let upperShadow(P,S) be the upper shadow of S in P, and similarly for lowerShadow.
+shadowHelper = (P,S,p) -> (
+    if not isSubset(S,P_*) then (
+        return "The given set is not a subset of the given poset";
+    );
+
+    shadow := {};
+
+    -- For each covering relation k={k#0,k#1} with k#0 in S, the element k#1 is supposed to be in the upper shadow of S.
+    -- For each covering relation k={k#0,k#1} with k#1 in S, the element k#0 is supposed to be in the lower shadow of S.
+    for k in coveringRelations(P) do (
+        if any(S, s->s===k#p) then (
+            shadow = append(shadow, k#(p-1));
```

Comment:
Using `append` inside a loop like this isn't efficient.  You could do something like:

```m2
for k in coveringRelations P list if any(S, s -> s === k#p) then k#(p - 1) else continue
```